File: bot-api/java/src/test/java/dev/robocode/tankroyale/botapi/graphics/PointTest.java
package dev.robocode.tankroyale.botapi.graphics;

import org.junit.jupiter.api.Test;
import static org.junit.jupiter.api.Assertions.*;

/**
 * Tests for the {@link Point} class.
 */
public class PointTest {

    @Test
    public void givenCoordinates_whenConstructingPoint_thenGettersReturnSameValues() {
        // Given
        double x = 10.5;
        double y = -5.25;

        // When
        Point point = new Point(x, y);

        // Then
        assertEquals(x, point.getX(), "X coordinate should match constructor value");
        assertEquals(y, point.getY(), "Y coordinate should match constructor value");
    }

    @Test
    public void givenPoints_whenComparingEquality_thenBehaveAsExpected() {
        // Given
        Point point1 = new Point(1.0, 2.0);
        Point point2 = new Point(1.0, 2.0);
        Point point3 = new Point(1.0, 3.0);
        Point point4 = new Point(3.0, 2.0);

        // Then
        assertEquals(point1, point2, "Equal points should be equal");
        assertNotEquals(point1, point3, "Points with different y coordinates should not be equal");
        assertNotEquals(point1, point4, "Points with different x coordinates should not be equal");
        assertNotEquals(null, point1, "Point should not equal null");
        assertNotEquals("Not a point", point1, "Point should not equal non-point objects");
    }

    @Test
    public void givenEqualPoints_whenHashCode_thenSameHashCode() {
        // Given
        Point point1 = new Point(1.0, 2.0);
        Point point2 = new Point(1.0, 2.0);

        // Then
        assertEquals(point1.hashCode(), point2.hashCode(), "Equal points should have the same hash code");
    }

    @Test
    public void givenPoint_whenToString_thenIncludesXAndYCoordinates() {
        // Given
        Point point = new Point(1.0, 2.0);

        // When
        String result = point.toString();

        // Then
        assertTrue(result.contains("x=1.0"), "toString should contain X coordinate");
        assertTrue(result.contains("y=2.0"), "toString should contain Y coordinate");
    }

    @Test
    public void givenPoint_whenCheckingReflexiveEquality_thenEqualToItself() {
        // Given
        Point point = new Point(1.0, 2.0);

        // Then
        assertEquals(point, point, "A point should be equal to itself");
    }
}

File: bot-api/wasm/src/wasmJsTest/kotlin/dev/robocode/tankroyale/botapi/graphics/PointTest.kt
package dev.robocode.tankroyale.botapi.graphics

import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertNotEquals
import kotlin.test.assertNotNull
import kotlin.test.assertTrue

/**
 * Tests for the Point class.
 */
class PointTest {

    @Test
    fun testConstructorAndGetters() {
        // Given
        val x = 10.5
        val y = -5.25

        // When
        val point = Point(x, y)

        // Then
        assertEquals(x, point.x, "X coordinate should match constructor value")
        assertEquals(y, point.y, "Y coordinate should match constructor value")
    }

    @Test
    fun testEquality() {
        // Given
        val point1 = Point(1.0, 2.0)
        val point2 = Point(1.0, 2.0)
        val point3 = Point(1.0, 3.0)
        val point4 = Point(3.0, 2.0)

        // Then
        assertEquals(point1, point2, "Equal points should be equal")
        assertNotEquals(point1, point3, "Points with different y coordinates should not be equal")
        assertNotEquals(point1, point4, "Points with different x coordinates should not be equal")
        assertNotNull(point1, "Point should not equal null")
    }

    @Test
    fun testEqualityWithExactValues() {
        // Given - Note: Since we now use exact equality, small differences will make points unequal
        val point1 = Point(1.0, 2.0)
        val point2 = Point(1.0, 2.0) // Exactly the same values
        val point3 = Point(1.0 + 1e-11, 2.0 - 1e-11) // Small differences - should now be unequal
        val point4 = Point(1.0 + 1e-9, 2.0) // Larger differences - should be unequal

        // Then
        assertEquals(point1, point2, "Points with exactly the same values should be equal")
        assertNotEquals(point1, point3, "Points with any difference should not be equal (exact comparison)")
        assertNotEquals(point1, point4, "Points with differences should not be equal")
    }

    @Test
    fun testHashCode() {
        // Given
        val point1 = Point(1.0, 2.0)
        val point2 = Point(1.0, 2.0)

        // Then
        assertEquals(point1.hashCode(), point2.hashCode(), "Equal points should have the same hash code")
    }

    @Test
    fun testToString() {
        // Given
        val point = Point(1.0, 2.0)

        // When
        val result = point.toString()

        // Then
        assertTrue(result.contains("X=1.0"), "toString should contain X coordinate")
        assertTrue(result.contains("Y=2.0"), "toString should contain Y coordinate")
    }

    @Test
    fun testReflexiveEquality() {
        // Given
        val point = Point(1.0, 2.0)

        // Then
        assertEquals(point, point, "A point should be equal to itself")
    }
}
